### PR TITLE
Add a simple Share a Draft feature

### PR DESCRIPTION
**Pull request by brandonpayton**
This is a WIP PR for a very simple Share a Draft feature.

This feature allows authors to create a single link to share a draft with friends prior to publishing.

draft-sharing-on

**NOTE:** We worked on a richer version of this feature under #16002. This is a separate PR for the sake of simplicity and clarity.

I'm maintaining the following checklists as a personal reminder and to reflect progress.

- [x] Create share-a-draft component
- [ ] Component unit tests
- [x] Actions
- [x] Reducer
- [x] Selectors
- [x] Connect component to redux
- [ ] Data-layer provider
- [ ] Query component
- [ ] Determine how to reflect loading state in the UI
- [x] Add copy link functionality
- [ ] Usage stats
- [ ] Review `translate` calls for opportunity to be clearer
- [ ] Review accessibility
- [ ] Review RTL support
- [x] Document vision for the REST API
- [ ] Reasonably test across supported browsers

Some cleanup that needs to occur before merging

- [ ] Add missing JSDoc
- [ ] Making naming more consistent. Currently, some things are named "share-a-draft" and others "draft-sharing". Choose one.

/cc @nb @dmsnell

### CODE REVIEW

File: client/post-editor/editor-share-a-draft/index.jsx [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+						'Enable draft sharing and we’ll create a public link to your draft, ' +
+							'so you can ask a trusted reader for feedback before you publish.',
+					) }
+				</p>
+				<FormToggle>
+					{ translate( 'Enable draft sharing' ) }
+				</FormToggle>
+				<FormTextInputWithAction
+					readonly
+					action={ translate( 'Copy', { context: 'verb: imperative' } ) }
```
Comment (a8ci18n) [inline client/post-editor/editor-share-a-draft/index.jsx]: Hi! I've found a possible matching string that has already been translated **33** times:
`translate( 'Copy', { context: 'verb'} )` *ES Score: 14*
See 1 additional suggestion in the [PR translation status page](https://translate.wordpress.com/localci/status/automattic/wp-calypso/add/simple-share-a-draft)

Help me improve these suggestions: react with :thumbsdown: if the suggestion doesn't make any sense, or with :thumbsup: if it's a particularly good one (even if not implemented).

:information_source: This string already exists with the following **contexts**:
- `null` (no context)
- `verb`

 Would it make sense to reuse one of the above?

**Comment by brandonpayton** [comment]
Tonight, I added the basic UI component and copying the link when the Copy button is clicked.

basic-component-and-copy-link

File: client/post-editor/editor-share-a-draft/index.jsx [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+ * Internal dependencies
+ */
+import Accordion from 'components/accordion';
+import FormToggle from 'components/forms/form-toggle';
+import FormTextInputWithAction from 'components/forms/form-text-input-with-action';
+
+class EditorShareADraft extends Component {
+	static propTypes = {
+		translate: PropTypes.func.isRequired,
+		// Expose copySelection as a prop to facilitate unit test
```
Comment (dmsnell) [inline client/post-editor/editor-share-a-draft/index.jsx]: :+1:

**Comment by brandonpayton** [comment]
I added actions, the `draftSharing` reducer and state slice, and a `getDraftSharing` selector.

I don't love calling the state slice `draftSharing` but prefer it over `shareADraft` because a name with two adjacent capital letters feels off to me.

**Comment by dmsnell** [comment]
Just a note here that we like to use more semantically-meaningful action types like `DRAFT_SHARE_ADD` rather than `DRAFT_SHARE_RECEIVE` because `_RECEIVE` is so passive and doesn't clearly indicate what should happen next, like "okay, now I've got a share, what do you want me to do with it?" Also it doesn't make much sense from the standpoint of entering in a command in the console dispatcher to `_RECEIVE` since nothing "came in"

**Comment by brandonpayton** [comment]
The intent for that action is simple to relay state retrieved from the data store. Maybe it should be called POST_SHARE_A_DRAFT_SET?

**Comment by brandonpayton** [comment]
On re-reading your comment, I think I may have misunderstood. Are you saying an _ADD action would be appropriate for populating state retrieved from the data store?

Thanks for pointing to the passive, vague nature of _RECEIVE. That's a good point.

**Comment by dmsnell** [comment]
> Are you saying an _ADD action would be appropriate for populating state retrieved from the data store?

Yeah I'm saying that `_ADD` is an appropriate name for when we are wanting to inject some data into the state tree regardless of where it came from.

**Comment by brandonpayton** [comment]
I roughed out the data-layer provider but haven't added tests yet. I feel like the provider code should be simpler but haven't had time to think about it.

This may be my last update with another team finishing the work.

**Comment by apeatling** [comment]
@brandonpayton this cool. Would it be possible to get this finished?

**Comment by brandonpayton** [comment]
@apeatling, I believe this is being taken over by Samus and am not sure of its current status.

@jblz, is there anything you can add here? Is there value to leaving this issue open?

**Comment by apeatling** [comment]
@jblz I'm closing this one up, if you decide to rebase, please reopen it.

**Pull request closed without merging**